### PR TITLE
[mono][aot] Simplify the dedup code.

### PR DESCRIPTION
**Pull request by vargaz**
Instead of passing state around in a MonoAotState structure, use a few global variables.

### CODE REVIEW

**Comment by vargaz** [comment]
So this is just a bit of refactoring to make the dedup code easier to understand.
It also fixes an issue where some gshared methods were not AOT-ed so its needed for dedup to work on wasm.

File: src/mono/mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -14198,121 +14196,6 @@ mono_read_method_cache (MonoAotCompile *acfg)
...
-		*astate = *global_aot_state;
-	}
-
-	if (!*astate) {
-		*astate = alloc_aot_state ();
-		*global_aot_state = *astate;
-	}
-
-	acfg->dedup_cache = (*astate)->cache;
-	acfg->dedup_stats = (*astate)->stats;
```
Comment (lambdageek) [inline src/mono/mono/mini/aot-compiler.c:14261]: `mono_dedup_log_stats` needs `MonoAotCompile:dedup_stats` to be non-NULL and nothing else initializes it.

Comment (vargaz) [reply src/mono/mono/mini/aot-compiler.c:14261]: Removed that as well.

File: src/mono/mono/mini/aot-compiler.c [hunk 2]
```diff
@@ -4371,10 +4331,14 @@ add_extra_method_full (MonoAotCompile *acfg, MonoMethod *method, gboolean prefer
 	}
 
 	if ((acfg->aot_opts.dedup || acfg->aot_opts.dedup_include) && mono_aot_can_dedup (method)) {
```
Comment (kotlarmilos) [inline src/mono/mono/mini/aot-compiler.c:4333]: There are flags that might be redundant:
 - `dedup` indicates dedup-skip phase
 - `dedup_include` indicates dedup-include phase
 -  `dedup_collect_only` indicates collecting of dedup-able methods in dedup-include
 - `dedup_emit_mode` indicates emitting of dedup-able methods in dedup-include phase

`dedup_collect_only` and `dedup_emit_mode` could be one flag that indicates collecting/emitting in dedup-include phase.

File: src/mono/mono/mini/aot-compiler.c [hunk 3]
```diff
@@ -4371,10 +4331,14 @@ add_extra_method_full (MonoAotCompile *acfg, MonoMethod *method, gboolean prefer
...
-		mono_dedup_cache_method (acfg, method);
-
-		if (!acfg->dedup_emit_mode)
+		if (acfg->aot_opts.dedup) {
```
Comment (kotlarmilos) [inline src/mono/mono/mini/aot-compiler.c:4334]: Parent `if` might not be needed if the nested `if`/`else if` statements are expanded with `&& mono_aot_can_dedup (method)`

File: src/mono/mono/mini/aot-compiler.c [hunk 4]
```diff
@@ -15067,6 +14723,53 @@ emit_aot_image (MonoAotCompile *acfg)
 	return 0;
 }
 
+int
+mono_aot_assemblies (MonoAssembly **assemblies, int nassemblies, guint32 jit_opts, const char *aot_options)
```
Comment (kotlarmilos) [inline src/mono/mono/mini/aot-compiler.c:14727]: `mono_aot_assemblies` LGTM

File: src/mono/mono/mini/aot-compiler.c [hunk 5]
```diff
@@ -15067,6 +14723,53 @@ emit_aot_image (MonoAotCompile *acfg)
...
+	MonoAotOptions aot_opts;
+
+	init_options (&aot_opts);
+	mono_aot_parse_options (aot_options, &aot_opts);
+	if (aot_opts.direct_extern_calls && !(aot_opts.llvm && aot_opts.static_link)) {
+		fprintf (stderr, "The 'direct-extern-calls' option requires the 'llvm' and 'static' options.\n");
+		return 1;
+	}
+
+	if (aot_opts.dedup_include) {
```
[the remark this answers is not in the document]
Comment (vargaz) [reply src/mono/mono/mini/aot-compiler.c:14739]: This is what the msbuild tasks pass today, and it's a bit easier to pass.

**Review by lambdageek — APPROVED** [review]

**Pull request merged**